RIOT-OS/RIOT
# Port of RIOT-OS to stm32mp1

Dear RIOTers,

### Contribution description

This is the first port of RIOT-OS to stm32mp1 and more precisely stm32mp157 which is a MPU with Arm Dual Cortex-A7 650 MHz, Arm Cortex-M4.
More information: https://www.st.com/en/microcontrollers-microprocessors/stm32mp157c.html

RIOT-OS firmware is flashed to the Cortex-M4 core.

Board STM32MP157C-DK2 is the only supported with a limited scope of features:
* gpio
* uart
* timers

Some drivers are exactly the same than for other stm32 families (UART), some aren't (GPIO) needing to modify common parts.

Only Engineering mode (BOOT0 off and BOOT2 on) has been tested.
This mode only enables the Cortex-M4 core.

### Testing procedure

A recent openocd with stm32mp15x support is needed:
* URL : https://git.code.sf.net/p/openocd/code
* commit: e8cfdd4a7208c4c2e7713eff46edecb4abbe37d6

On host computer:
`make BOARD=stm32mp157c-dk2 -C tests/periph_gpio/ flash`

Once done, on target, you should get the RIOT shell and you can play with LEDs (PA14 here, green LED):
```
helpmain(): This is RIOT! (Version: 2020.10-devel-522-g61898-mp1_dev)
GPIO peripheral driver test

In this test, pins are specified by integer port and pin numbers.
So if your platform has a pin PA01, it will be port=0 and pin=1,
PC14 would be port=2 and pin=14 etc.

NOTE: make sure the values you use exist on your platform! The
      behavior for not existing ports/pins is not defined!
> init_out 0 14
> clear 0 14
> set 0 14
> clear 0 14
```

The main challenge here is that Cortex-M4 core has no flash memory to boot on.
Thus bootloaders, or classic flashing process could not be applied.
Firmware has to be loaded manually on each reboot with the previous command.
The trick is to consider part of the RAM as ROM and load the firmware into that pseudo-ROM.
Next steps are:
* Load the RIOT firmware from Linux
* Load the RIOT firmware from u-boot
* Create communication between Linux and RIOT-OS using IPCC (https://www.kernel.org/doc/Documentation/devicetree/bindings/mailbox/stm32-ipcc.txt)
* Adding other peripheral drivers (i2c, spi, qdec, etc...)

### Issues/PRs references

None
